### PR TITLE
Implement rollback farm for mergeTree client

### PR DESCRIPTION
**Pull request by scarlettjlee**
## Description
Add tests that will randomly make edits and sometimes roll them back.

There are two bugs uncovered so far:
- previousProps for annotations don't get updated when the segment[s] are split afterward, so rolling back that annotation fails because the number of previousProps don't match the new segment count. - FIXED (in separate, completed PR)
- the cachedLength (3) is somehow mismatched to the content length (2). There are a LOT of removed segments at this point. This throws when it tries to insert text at a location that is past the end, not during rollback.
- properties don't match on the rollback client. This broke sometime after this PR was started. :(

## Notes for reviewers

Should zamboni be validated?

### CODE REVIEW

**Comment by anthony-murphy** [comment]
this seems to duplicate quite of bit of the existing code. i would probably model is more like: packages\dds\merge-tree\src\test\client.localReferenceFarm.spec.ts

i think the flow would look something like:

1. create clients
2. loop rounds (rounds generally mean reused/dirty clients), probably like 10 round
3. preform random actions, maybe like 10 ops, on clients with the existing generateOperationMessagesForClients
4. ack those random ops, the goal is just to get funky tree shapes with removed segments and stuff
5. on a single client, i would probably pick client 1/B call generateOperationMessagesForClients, this i would do as a range from 1 to 32 or something, 
6. rollback all the ops done on client  1/B
7.  preform random actions, maybe like 10 ops, on clients with the existing generateOperationMessagesForClients
8. ack those random ops, the goal is to ensure that client B is left in a good state

File: packages/dds/merge-tree/src/test/client.rollbackFarm.spec.ts [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+    opsPerRollbackRange: { min: 1, max: 15 },
+    opsPerRoundRange: { min: 1, max: 100 },
+    rounds: 5,
+    operations: allOperations,
+    incrementalLog: true,
+    growthFunc: (input: number) => input + 1,
+};
+
+export const defaultOptions: IRollbackFarmConfig = {
+    minLength: { min: 1, max: 512 },
```
Comment (anthony-murphy) [inline packages/dds/merge-tree/src/test/client.rollbackFarm.spec.ts:65]: no need to go so big. generally, the most interesting cases are small length, so lots of conflicts on position, i'd probably make max 32 for something like this. big enough to get deepish trees, but small enough to run fast. we never fine bugs in the high length of conflict farm, so it more there for sanity

File: packages/dds/merge-tree/src/test/client.rollbackFarm.spec.ts [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+    rollbackOnlyClients: { min: 1, max: 1 },
```
Comment (anthony-murphy) [inline packages/dds/merge-tree/src/test/client.rollbackFarm.spec.ts:66]: i don't think this need to be part of config, just always have 1

File: packages/dds/merge-tree/src/test/client.rollbackFarm.spec.ts [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+    editOnlyClients: { min: 1, max: 1 },
```
Comment (anthony-murphy) [inline packages/dds/merge-tree/src/test/client.rollbackFarm.spec.ts:67]: i don't think this needs to be explict, 0 is read-only, 1 is rollback, and then a couple editors

File: packages/dds/merge-tree/src/test/client.rollbackFarm.spec.ts [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+    rollbackAndEditClients: { min: 1, max: 3 },
+    opsPerRollbackRange: { min: 1, max: 128 },
+    opsPerRoundRange: { min: 1, max: 128 },
```
Comment (anthony-murphy) [inline packages/dds/merge-tree/src/test/client.rollbackFarm.spec.ts:70]: i would probably just go with a fixed number of ops, like 10, for non-rollback, this times rounds will represent the ops on any merge tree, so will still get lots of coverage

File: packages/dds/merge-tree/src/test/client.rollbackFarm.spec.ts [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+                    clients,
+                    minLength,
+                    opts,
+                );
+
+                if (clients.clients.length === maxClientCount) {
+                    break;
+                }
+
+                clients.clients.forEach((c) => c.updateMinSeq(seq));
```
Comment (anthony-murphy) [inline packages/dds/merge-tree/src/test/client.rollbackFarm.spec.ts:117]: if you run this between rounds zamboni will run, in addition to running on remote ops and acts. beyond that you don't need to do anything specific

**Comment by scarlettjlee** [comment]
I was trying to fix the failures first before completing this pull request, but considering that another bug has crept in in the meantime, I'd like to get this checked in to prevent more regressions. So I've limited the functionality to the parts that work. The broken parts can be re-enabled as the bugs are fixed.

**Review by anthony-murphy — APPROVED** [review]

**Pull request merged**